File: 프로그래머스/4/12929. 올바른 괄호의 갯수/올바른 괄호의 갯수.js
function solution(n) {
    let answer = new Array(n+1).fill(0);
    answer[0]=1;
    answer[1]=1;
    
    for(let i=2; i<n+1; i++){
        for(let a=0; a<=i-1; a++){
            let b=i-1-a;
            answer[i]+=answer[a]*answer[b];
        }
    }
    
    return answer[n];
}

File: 프로그래머스/4/12929. 올바른 괄호의 갯수/올바른 괄호의 갯수.java
class Solution {
    public int solution(int n) {
        int[] dp = new int[n+1];
        dp[0] = 1;
        dp[1] = 1;
        
        // (a쌍)b쌍 형식, a,b는 0개일 수도 있다. 카탈란수.
        for(int i=2; i<=n; i++){ // i=괄호 쌍
            int total = i-1; // 고정한 한 쌍의 괄호 제외 나머지 괄호 개수
            for(int a=0; a<=total; a++){
                int b = total-a;
                dp[i] += dp[a]*dp[b]; 
            }
        }
        
        return dp[n];
    }
}